lib/tree.rb

```diff
@@ -16,49 +16,133 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n)
+  # Space Complexity: ??
+  def add(key, value = nil)
```

Comment:
👍 The space complexity is O(log n) if the tree is balanced and O(n) if it's not.  Same with time complexity.